### PR TITLE
UniswapV3 fee support in price calculation

### PR DESCRIPTION
**Pull request by WalidOfNow**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/misc/UniswapV3OracleWrapper.sol [hunk 1]
```diff
@@ -26,87 +28,251 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
...
+                    feeGrowthGlobal0X128 -
+                    feeParams.feeGrowthOutside0X128Upper;
+                feeGrowthAbove1X128 =
+                    feeGrowthGlobal1X128 -
+                    feeParams.feeGrowthOutside1X128Upper;
+            }
+            uint256 feeGrowthInside0X128;
+            uint256 feeGrowthInside1X128;
+
+            feeGrowthInside0X128 =
```
Comment (GopherJ) [inline contracts/misc/UniswapV3OracleWrapper.sol:255]: shall we use `Tick.getFeeGrowthInside` instead?

File: contracts/misc/UniswapV3OracleWrapper.sol [hunk 2]
```diff
@@ -26,87 +28,251 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
...
+                feeGrowthGlobal0X128 -
+                feeGrowthBelow0X128 -
+                feeGrowthAbove0X128;
+            feeGrowthInside1X128 =
+                feeGrowthGlobal1X128 -
+                feeGrowthBelow1X128 -
+                feeGrowthAbove1X128;
+
+            unclaimedTokenAFees =
```
Comment (GopherJ) [inline contracts/misc/UniswapV3OracleWrapper.sol:264]: we can use `FullMath.mulDiv` maybe to be more accurate

File: contracts/misc/UniswapV3OracleWrapper.sol (hunk 1 above)
Comment (zhoujia6139) [reply contracts/misc/UniswapV3OracleWrapper.sol:255]: which `getFeeGrowthInside`?

File: contracts/misc/UniswapV3OracleWrapper.sol (hunk 2 above)
Comment (zhoujia6139) [reply contracts/misc/UniswapV3OracleWrapper.sol:264]: yeah. i think so.

File: contracts/misc/UniswapV3OracleWrapper.sol [hunk 3]
```diff
@@ -141,4 +174,221 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
...
+
+    function _getLiquidityAmount(
+        uint160 sqrtRatioX96,
+        UinswapV3PositionData memory positionData
+    ) internal pure returns (uint256 token0Amount, uint256 token1Amount) {
+        (token0Amount, token1Amount) = LiquidityAmounts.getAmountsForLiquidity(
+            sqrtRatioX96,
+            TickMath.getSqrtRatioAtTick(positionData.tickLower),
+            TickMath.getSqrtRatioAtTick(positionData.tickUpper),
+            positionData.liquidity
```
Comment (GopherJ) [inline contracts/misc/UniswapV3OracleWrapper.sol:284]: this is the virtual liquidity which cannot reflect users' position, I think we should use `SqrtPriceMath.getAmount0Delta` instead

File: contracts/misc/UniswapV3OracleWrapper.sol (hunk 1 above)
Comment (GopherJ) [reply contracts/misc/UniswapV3OracleWrapper.sol:372]: in v3-core/contracts/libraries/Tick.sol there is `getFeeGrowthInside` function, we can also use this `Tick` library

File: contracts/misc/UniswapV3OracleWrapper.sol [hunk 4]
```diff
@@ -141,4 +174,221 @@ contract UniswapV3OracleWrapper is IAtomicPriceAggregator {
...
+        oracleData.tokenBPrice = PARASPACE_ORACLE.getAssetPrice(
+            positionData.tokenB
+        );
+
+        oracleData.tokenADecimal = IERC20Detailed(positionData.tokenA)
+            .decimals();
+        oracleData.tokenBDecimal = IERC20Detailed(positionData.tokenB)
+            .decimals();
+
+        // TODO using bit shifting for the 2^96
```
Comment (GopherJ) [inline contracts/misc/UniswapV3OracleWrapper.sol:196]: we can use `x << FixedPoint96.RESOLUTION`

**Review by GopherJ — APPROVED** [review]

File: contracts/misc/UniswapV3OracleWrapper.sol (hunk 1 above)
Comment (zhoujia6139) [reply contracts/misc/UniswapV3OracleWrapper.sol:372]: we don't have enough parameter info for this function.

**Pull request merged**